In [39]:
import re
from test_utilities import run_tests_params

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 5: Doesn't He Have Intern-Elves For This? ---</h2><p>Santa needs help figuring out which strings in his text file are naughty or nice.</p>
<p>A <em>nice string</em> is one with all of the following properties:</p>
<ul>
<li>It contains at least three vowels (<code>aeiou</code> only), like <code>aei</code>, <code>xazegov</code>, or <code title="John Madden John Madden John Madden">aeiouaeiouaeiou</code>.</li>
<li>It contains at least one letter that appears twice in a row, like <code>xx</code>, <code>abcdde</code> (<code>dd</code>), or <code>aabbccdd</code> (<code>aa</code>, <code>bb</code>, <code>cc</code>, or <code>dd</code>).</li>
<li>It does <em>not</em> contain the strings <code>ab</code>, <code>cd</code>, <code>pq</code>, or <code>xy</code>, even if they are part of one of the other requirements.</li>
</ul>
<p>For example:</p>
<ul>
<li><code>ugknbfddgicrmopn</code> is nice because it has at least three vowels (<code>u...i...o...</code>), a double letter (<code>...dd...</code>), and none of the disallowed substrings.</li>
<li><code>aaa</code> is nice because it has at least three vowels and a double letter, even though the letters used by different rules overlap.</li>
<li><code>jchzalrnumimnmhp</code> is naughty because it has no double letter.</li>
<li><code>haegwjzuvuyypxyu</code> is naughty because it contains the string <code>xy</code>.</li>
<li><code>dvszwmarrgswjxmb</code> is naughty because it contains only one vowel.</li>
</ul>
<p>How many strings are nice?</p>
</article>


In [40]:
tests = [
    {
        "name": "Example 1",
        "s": "ugknbfddgicrmopn",
        "expected": True,
    },
    {
        "name": "Example 2",
        "s": "aaa",
        "expected": True,
    },
    {
        "name": "Example 3",
        "s": "jchzalrnumimnmhp",
        "expected": False,
    },
    {
        "name": "Example 4",
        "s": "haegwjzuvuyypxyu",
        "expected": False,
    },
    {
        "name": "Example 5",
        "s": "dvszwmarrgswjxmb",
        "expected": False,
    },
]


def nice(s: str) -> bool:
    # It contains at least three vowels (aeiou only)
    vowels = r"[a|e|i|o|u]"
    if len(re.findall(vowels, s)) < 3:
        return False
    # It contains at least one letter that appears twice in a row
    double_vowels = r"(.)\1"
    if re.search(double_vowels, s) is None:
        return False
    # It does not contain the strings ab, cd, pq, or xy
    illegal = r"ab|cd|pq|xy"
    return re.search(illegal, s) is None


run_tests_params(nice, tests)


Test Example 1 passed, for nice.
Test Example 2 passed, for nice.
Test Example 3 passed, for nice.
Test Example 4 passed, for nice.
Test Example 5 passed, for nice.
Success


In [41]:
s = """
ugknbfddgicrmopn
aaa
jchzalrnumimnmhp
haegwjzuvuyypxyu
dvszwmarrgswjxmb
"""


def count_nice(s: str) -> int:
    return sum(1 for l in s.strip().splitlines() if nice(l))


assert count_nice(s) == 2

In [42]:
with open("../input/day5.txt") as f:
    print(count_nice(f.read()))

238


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>238</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Realizing the error of his ways, Santa has switched to a better model of determining whether a string is naughty or nice.  None of the old rules apply, as they are all clearly ridiculous.</p>
<p>Now, a nice string is one with all of the following properties:</p>
<ul>
<li>It contains a pair of any two letters that appears at least twice in the string without overlapping, like <code>xyxy</code> (<code>xy</code>) or <code>aabcdefgaa</code> (<code>aa</code>), but not like <code>aaa</code> (<code>aa</code>, but it overlaps).</li>
<li>It contains at least one letter which repeats with exactly one letter between them, like <code>xyx</code>, <code>abcdefeghi</code> (<code>efe</code>), or even <code>aaa</code>.</li>
</ul>
<p>For example:</p>
<ul>
<li><code>qjhvhtzxzqqjkmpb</code> is nice because is has a pair that appears twice (<code>qj</code>) and a letter that repeats with exactly one letter between them (<code>zxz</code>).</li>
<li><code>xxyxx</code> is nice because it has a pair that appears twice and a letter that repeats with one between, even though the letters used by each rule overlap.</li>
<li><code>uurcxstgmygtbstg</code> is naughty because it has a pair (<code>tg</code>) but no repeat with a single letter between them.</li>
<li><code>ieodomkazucvgmuy</code> is naughty because it has a repeating letter with one between (<code>odo</code>), but no pair that appears twice.</li>
</ul>
<p>How many strings are nice under these new rules?</p>
</article>
</main>


In [43]:
from collections import defaultdict


tests = [
    {
        "name": "Example 1",
        "s": "qjhvhtzxzqqjkmpb",
        "expected": True,
    },
    {
        "name": "Example 2",
        "s": "xxyxx",
        "expected": True,
    },
    {
        "name": "Example 3",
        "s": "uurcxstgmygtbstg",
        "expected": False,
    },
    {
        "name": "Example 4",
        "s": "ieodomkazucvgmuy",
        "expected": False,
    },
]


def at_least_one_non_overlapping_pair(s: str) -> bool:
    indici = defaultdict(list)
    for i in range(1, len(s)):
        double = s[i - 1 : i + 1]
        if double in indici:
            for ii in indici[double]:
                if i - ii > 1:
                    return True
        indici[double].append(i)
    return False


def at_least_one_repeating_letter_with_one_letter_in_between(
    s: str,
) -> bool:
    for i in range(2, len(s)):
        if s[i - 2] == s[i]:
            return True
    return False


def nice_II(s: str) -> bool:
    # It contains a pair of any two letters that appears at least
    # twice in the string without overlapping,
    # It contains at least one letter which repeats with exactly one letter between them
    return at_least_one_repeating_letter_with_one_letter_in_between(
        s
    ) and at_least_one_non_overlapping_pair(s)


run_tests_params(nice_II, tests)


Test Example 1 passed, for nice_II.
Test Example 2 passed, for nice_II.
Test Example 3 passed, for nice_II.
Test Example 4 passed, for nice_II.
Success


In [44]:
s = """
qjhvhtzxzqqjkmpb
xxyxx
uurcxstgmygtbstg
ieodomkazucvgmuy
"""


def count_nice_II(s: str) -> int:
    return sum(1 for l in s.strip().splitlines() if nice_II(l))


assert count_nice_II(s) == 2

In [45]:
with open("../input/day5.txt") as f:
    print(count_nice_II(f.read()))

69


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>69</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2015">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="5/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22Doesn%27t+He+Have+Intern%2DElves+For+This%3F%22+%2D+Day+5+%2D+Advent+of+Code+2015&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F5&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22Doesn%27t+He+Have+Intern%2DElves+For+This%3F%22+%2D+Day+5+%2D+Advent+of+Code+2015+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F5';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
